src/librustc_mir/const_eval/machine.rs

```diff
@@ -86,12 +86,15 @@ impl<'mir, 'tcx> InterpCx<'mir, 'tcx, CompileTimeInterpreter> {
 }
 
 /// Extra machine state for CTFE, and the Machine instance
-pub struct CompileTimeInterpreter {
+pub struct CompileTimeInterpreter<'mir, 'tcx> {
     /// For now, the number of terminators that can be evaluated before we throw a resource
     /// exhuastion error.
     ///
     /// Setting this to `0` disables the limit and allows the interpreter to run forever.
     pub steps_remaining: usize,
+
+    /// The virtual call stack.
+    pub(crate) stack: Vec<Frame<'mir, 'tcx, (), ()>>,
```

Comment:
The code duplication between the CTFE machine and the const-prop machine is started to get a little too much here. Could you find some way to share common code between them?

Reply:
Since the machine structs have only 1-2 fields, I am not sure whether we can share them efficiently. However, I think i could define trait `CommonMachine` that would contain the implementations of all associate types and methods that are identical for CTFE and const-prop machines. Does this sound reasonable to you?

Reply:
That would help for everything else but not the stack then? Hm. This is where inheritance would be useful. ;)

If it doesn't actually help for the stack then yes this sounds reasonable but it should probably be done separately. Could you open an issue to track that?

Reply:
> This is where inheritance would be useful. ;)

Yes...

> Could you open an issue to track that?

Opened: https://github.com/rust-lang/rust/issues/71129

If it looks good to you, you could maybe even label it as a *good first issue* or whatever the label is used for marking issues that are good for newcomers.